Add _foreach_add_(TensorList tensors, Scalar scalar) API

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43727 WIP asgd
* #43726 Rewrote rprop optimizer with _foreach APIs
* #43725 Rewrote rmsprop with _forerach APIs
* #43724 Rewrote SGD with _foreach APIs
* #43562 Rewrote AdamW with _foreach APIs
* #43507 [WIP] Rewrote adam optimizer with foreach APIs
* #43413 WIP. Added alpha overloads for add/sub ops with lists
* #42760 [IGNORE] [WIP] Added pointwise ops: addcdiv, addcmul
* #42537 Add unary ops: exp and sqrt
* #42536 Add binary ops for _foreach APIs
* #42533 Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs
* **#42531 Add _foreach_add_(TensorList tensors, Scalar scalar) API**

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

---------------
**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331892](https://our.internmc.facebook.com/intern/diff/D23331892)

Commit:
Added add_scalar_

[ghstack-poisoned]

Commit:
Update on "[IGNORE] Added add_scalar_"




[ghstack-poisoned]

Commit:
Update on "[IGNORE] Added add_scalar_"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API

**Tests**
Tested via unit tests

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[IGNORE] Added add_scalar_"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API

**Tests**
Tested via unit tests

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "Add _foreach_add_(TensorList tensors, Scalar scalar) "


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "Add _foreach_add_(TensorList tensors, Scalar scalar) API"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

ngimel reviewed (commented):
Can you briefly describe why changes to autograd and jit are necessary?

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5450,6 +5450,13 @@
+  device_guard: False  
+  variants: function
+  dispatch:
+    CPU: foreach_add_scalar__kernel_fallback
```
please put the underscore at the end of the name, i.e. `foreach_add_scalar_kernel_fallback_`, that's much easier to read.

izdeby replied:
Done

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5450,6 +5450,13 @@
     CPU: foreach_add_scalar_kernel_fallback
     CUDA: foreach_tensor_add_scalar_kernel_cuda
 
+- func: _foreach_add_.Scalar(Tensor[](a!) self, Scalar scalar) -> Tensor[](a!)
```
we don't have in-place functions, only methods.  Is it feasible to make this an out-variant function?

gchanan replied:
actually this isn't quite true, a bunch of stuff in nn::functional is defined this way (although they aren't typically directly accessible).

gchanan replied:
sounds good then!

gchanan replied:
wouldn't `Tensor[](a!)` mean that the list is being mutated, and `Tensor(a!)[]` be a list of mutable tensors?

@suo is that right?  Is this safe to add?

ezyang replied:
Coming in here late, I noticed this problem too while working on the new codegen at https://github.com/pytorch/pytorch/pull/42629 I also agree that `Tensor(a!)[]` is the correct type, and what was landed (`Tensor[](a!)`) is not correct.

ezyang replied:
I added a small fix for the problem in the new codegen. Besides

```
- - func: _foreach_add_.Scalar(Tensor[](a!) self, Scalar scalar) -> ()
-   device_guard: False  
+ - func: _foreach_add_.Scalar(Tensor(a!)[] self, Scalar scalar) -> ()
+   device_guard: False
```

I also had to add a special case to make void return for trailing underscore ok:

```

        if self.name.name.inplace:
            # TODO: fixme
            if str(self.name) not in [
                    '_amp_non_finite_check_and_unscale_',
                    '_foreach_add_.Scalar']:
                assert len(self.returns) == 1
```

gchanan commented on tools/autograd/gen_variable_type.py:
```diff
@@ -139,6 +139,8 @@
     # Functions that return integers should not have output that require gradients
     'argmax', 'argmin', 'argsort',
+    # Foreach functions for now.
+    '_foreach_add_',
```
why?

gchanan commented on torch/csrc/jit/frontend/tracer.h:
```diff
@@ -355,6 +359,7 @@ TORCH_API void addOutput(
     Node* node,
     const c10::intrusive_ptr<c10::ivalue::Object>& output);
 
+TORCH_API void addOutput(Node* node, const at::TensorList outputs);
```
why is this not captured by:
`TORCH_API void addOutput(Node* node, const std::vector<at::Tensor>& list);` ?

izdeby replied:
Because TensorList is not a std::vector but an [ArrayRef](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/templates/TensorBody.h#L50)

in tracer.cpp im passing the given TensorList to this overload
```
void addOutput(Node* node, const at::TensorList outputs) {
  return addOutput(node, outputs.vec());
}
```

gchanan replied:
how can you return an ArrayRef?

gchanan replied:
I guess we know that the ArrayRef is valid through the entire path?

gchanan replied:
I was actually making a different point.

The C++ signature is:
`std::vector<Tensor> foreach_add_scalar_kernel_fallback_(TensorList tensors, Scalar scalar) {`

which suggests a vector is an output, but here you need a TensorList as an output?

gchanan replied:
As @ezyang points out, we don't need method chaining (we don't support it on TensorLists) and we do have mutable signatures that don't have returns:
https://github.com/pytorch/pytorch/blob/e8f4b04d9ae1880fd8d09dde40639e86a0d2c585/aten/src/ATen/native/native_functions.yaml#L1006

You might have to play around with some combinations to get the codegen correct, though.

izdeby replied:
Done!

izdeby commented on aten/src/ATen/native/ForeachOpsKernels.cpp:
```diff
@@ -5,10 +5,21 @@ std::vector<Tensor> foreach_add_scalar_kernel_fallback(TensorList tensors, Scala
   std::vector<Tensor> result;
-  for (int i = 0; i < tensors.size(); i++) {
-    auto temp = tensors[i].add(scalar);
+  for (auto& t : tensors) {
```
Done

izdeby commented on test/test_foreach.py:
```diff
@@ -1,20 +1,34 @@
+        N = 20
+        H = 20
+        W = 20
+        tensors = [torch.zeros(H, W, device=device, dtype=dtype) for n in range(N)]
```
Done

gchanan reviewed (commented):
what do you expect to happen when the inplace fails?  Is the entire thing committed or not?

```
>>> x=torch.ones(2,3, dtype=torch.float, device='cuda:1')
>>> y=torch.ones(2,3, dtype=torch.long, device='cuda:1')
>>> torch._foreach_add_([x,y], 3.5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: result type Float can't be cast to the desired output type Long
>>> x
tensor([[4.5000, 4.5000, 4.5000],
        [4.5000, 4.5000, 4.5000]], device='cuda:1')
>>> y
tensor([[1, 1, 1],
        [1, 1, 1]], device='cuda:1')

gchanan commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -88,6 +88,11 @@ inline void increment_version(Tensor & t) {
   impl::bump_version(t);
 }
 
+inline void increment_version(at::TensorList tensors) {
```
so won't the version counter be different depending on if we take the fast path or not?

izdeby replied:
its gone

Commit:
Update on "[WIP] Add _foreach_add_(TensorList tensors, Scalar scalar) API"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP] Add _foreach_add_(TensorList tensors, Scalar scalar) API"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP] Add _foreach_add_(TensorList tensors, Scalar scalar) API"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5467,6 +5467,13 @@
+  variants: function
+  dispatch:
+    CPU: foreach_add_scalar_kernel_fallback_
+    CUDA: foreach_tensor_add_scalar_kernel_cuda_
```
these having different names is really confusing, it sounds like the CPU variant doesn't involve tensors.

gchanan replied:
Also using fallback is confusing because a fallback kernel is a different concept entirely.  I'd suggest using "slow"

izdeby replied:
Renamed
```
CPU: foreach_tensor_add_scalar_kernel_slow_
CUDA: foreach_tensor_add_scalar_kernel_cuda_
```

Commit:
Update on "[WIP] Add _foreach_add_(TensorList tensors, Scalar scalar) API"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP] Add _foreach_add_(TensorList tensors, Scalar scalar) API"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP] Add _foreach_add_(TensorList tensors, Scalar scalar) API"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP] Add _foreach_add_(TensorList tensors, Scalar scalar) API"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

---------------
**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

[ghstack-poisoned]

izdeby commented:
@gchanan, @ngimel, @cpuhrsch 
Updated the stack according to all the discussions. This PR is a bit messy but all the next ones should be straightforward. 
Changes since last review: 
- Renamed `foreach_add_scalar_kernel_fallback` ---> `foreach_tensor_add_scalar_kernel_slow_`
- updated tests to cover 4 different scalars 
- in place ops now return void 
- Moved `AddScalarFunctor_` and `AddScalarFunctor` to _native/cuda/ForeachFunctors.cuh_ which will be used for all functors in future. 
- _native/cuda/ForeachUtils.cuh_ --> _native/ForeachUtils.h_

Commit:
Update on "Add _foreach_add_(TensorList tensors, Scalar scalar) API"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

---------------
**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

[ghstack-poisoned]

Commit:
Update on "Add _foreach_add_(TensorList tensors, Scalar scalar) API"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

---------------
**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

[ghstack-poisoned]

Commit:
Update on "Add _foreach_add_(TensorList tensors, Scalar scalar) API"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

---------------
**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331892](https://our.internmc.facebook.com/intern/diff/D23331892)

[ghstack-poisoned]

Commit:
Update on "Add _foreach_add_(TensorList tensors, Scalar scalar) API"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

---------------
**In this PR**
- Adding a `std::vector<Tensor> _foreach_add_(TensorList tensors, Scalar scalar)` API
- Resolving some additional comments from previous [PR](https://github.com/pytorch/pytorch/pull/41554).

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331892](https://our.internmc.facebook.com/intern/diff/D23331892)

[ghstack-poisoned]
